### PR TITLE
Revert "Remove unnecessary trait accessibility check."

### PR DESCRIPTION
**Pull request by alexcrichton**
These commits revert https://github.com/rust-lang/rust/pull/28504 and add a regression test pointed out by @petrochenkov, it's not immediately clear with the regression that the accessibility check should be removed, so for now preserve the behavior on stable by default.

r? @nrc

### CODE REVIEW

File: src/test/compile-fail/privacy-ufcs.rs [hunk 1]
```diff
@@ -1,24 +0,0 @@
...
-        fn baz() {}
-    }
-
-    impl Bar for i32 {}
-}
-
-fn main() {
-    <i32 as ::foo::Bar>::baz(); //~ERROR method `baz` is inaccessible
-                                //~^NOTE: trait `Bar` is private
-}
```
Comment (nrc) [inline src/test/compile-fail/privacy-ufcs.rs]: Do we need to remove this test - it should still fail, right?

**Comment by nrc** [comment]
r+ with or without test

File: src/test/compile-fail/privacy-ufcs.rs (hunk 1 above)
Comment (alexcrichton) [inline src/test/compile-fail/privacy-ufcs.rs]: Oh true, this was just lumped in the `git revert` commands, but I'll throw it in regardless

**Pull request merged**